### ISSUE TITLE
Nano - regerror: assertion "errbuf != NULL" failed

### ISSUE THREAD

**Issue by 4oo4** [termux/termux-packages#877]
Just started getting this error when trying to launch nano (v2.7.5) yesterday. The full error it's giving is:

`bionic/libc/upstream-netbsd/lib/libc/regex/regerror.c:175: regerror assertion "errbuf !=NULL" failed`

This is only on my Samsung Galaxy Tab running Cyanogenmod 13 (it's an unofficial CM but based off of CM stable), my Nexus 5x running LineageOS 14.1 works normally. I know I used nano sometime in the last week up until I started getting this.

Let me know if there are any logs I can provide.

**Comment by vishalbiswas** [comment]
This happens in the first call to regerror to get the size of the errbuf required. The manpage states that errbuf is ignored if last argument is 0, but this isn't exactly the case in bionic's (netbsd's) regex implemention. errbuf is asserted for NULL unconditionally.

**Comment by fornwall** [comment]
> This happens in the first call to regerror to get the size of the errbuf required. The manpage states that errbuf is ignored if last argument is 0, but this isn't exactly the case in bionic's (netbsd's) regex implemention. errbuf is asserted for NULL unconditionally.

Nice find (for reference: [regerror.c in bionic from netbsd](https://android.googlesource.com/platform/bionic/+/master/libc/upstream-netbsd/lib/libc/regex/regerror.c#162))! So it seems that hitting any call to `regerror` in nano should trigger the problem, as nano always seems to pass NULL for the error buffer in the initial call:

```sh
$ ag regerror
src/rcfile.c
264:	size_t len = regerror(rc, &preg, NULL, 0);
267:	regerror(rc, &preg, str, len);

src/search.c
51:	size_t len = regerror(value, &search_regexp, NULL, 0);
54:	regerror(value, &search_regexp, str, len);

src/nano.c
2435:	size_t size = regerror(quoterc, &quotereg, NULL, 0);
2438:	regerror(quoterc, &quotereg, quoteerr, size);
```

**Comment by fornwall** [comment]
Hm, the assertion made through the `_DIAGASSERT` macro seems to be [disabled in bionic](https://android.googlesource.com/platform/bionic/+/master/libc/upstream-netbsd/android/include/netbsd-compat.h#23) due to these kind of problems:

```c
// NetBSD uses _DIAGASSERT to null-check arguments and the like,
// but it's clear from the number of mistakes in their assertions
// that they don't actually test or ship with this.
#define _DIAGASSERT(e) /* nothing */
```

This was [committed in 2016-02-13](https://android.googlesource.com/platform/bionic/+/8739a1a513b3d5c3043d764d2719b6a5e66ad21d%5E1..8739a1a513b3d5c3043d764d2719b6a5e66ad21d/) for exactly this problem: `Fix regerror(..., nullptr, 0).` This should mean that it was fixed in nougat, which explains why the error doesn't happen in LineageOS 14.1 (based on nougat) but in Cyanogenmod 13 (based on marshmallow).

**Comment by fornwall** [comment]
@4oo4 Nano has been updated to version 2.8.0 - can you run `packages upgrade` and check if the updated package fixes the problem?

**Comment by 4oo4** [comment]
@fornwall I was actually about to let you know that the upgrade to 2.8.0 fixed it. Thanks for the update!

**Comment by derek1906** [comment]
Recently I'm getting the exact error with `grep`:

```
ps ax | grep '\{'
bionic/libc/upstream-netbsd/lib/libc/regex/regerror.c:175: regerror: assertion "errbuf != NULL" failed
```

**Comment by fornwall** [comment]
@derek1906 

- Which Android version are you running?
- Can you update all packages (`pkg up`) and verify that the problem remains?

**Comment by derek1906** [comment]
I'm using Termux on Android 6.0.1 (Samsung Note 4) and after running `pkg up` the error is still popping up (instead of showing the expected message `grep: braces not balanced`). Not sure what the exact cause is but the error message matches the one shown in this thread.

**Comment by fornwall** [comment]
The root cause is the one mentioned above, that regerror() does not work correctly in bionic until Android 7.0, and grep would need to be patched to work around it.

**Comment by Deathlive** [comment]
@fornwall Is it possible to fix this problem by Android 5.1 Lolipop?
